File: 238-product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int productOfAllExcept_0 = 1;
        int zeroCount = 0;
        for (int num : nums) {
            if (num == 0)
                zeroCount++;
            else
                productOfAllExcept_0 *= num;
        }

        int n = nums.size();
        vector<int> ans(n, 0);

        if (zeroCount > 1)
            return ans; // all will have zero

        for (int i = 0; i < n; i++) {
            if (zeroCount == 0) {
                // there are no zeroes
                ans[i] = productOfAllExcept_0 / nums[i];
            } else {
                // There is 1 zero
                if (nums[i] == 0) {
                    ans[i] = productOfAllExcept_0;
                } else {
                    ans[i] = 0;
                }
            }
        }

        return ans;
    }
};

File: 238-product-of-array-except-self/product-of-array-except-self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function(nums) {
    const prefixArr = [];

    let product = 1;
    for(let i=0; i<nums.length; i++) {
        const num = nums[i];
        prefixArr[i] = product;
        product *= num;
    }

    const result = [];
    product = 1;
    for(let i=nums.length - 1; i>=0; i--) {
        const num = nums[i];
        const postfixValAtI = product;
        result[i] = prefixArr[i] * postfixValAtI;
        product *= num;
    }

    return result;
};